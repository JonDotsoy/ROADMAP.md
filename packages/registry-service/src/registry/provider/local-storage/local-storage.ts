import type { StorageProviderDTO } from "../../dtos/storage-provider-dto";
import * as fs from "fs/promises";

// const storageLocation = new URL("./.storage/", import.meta.url);

// await fs.mkdir(storageLocation, { recursive: true });

const sha256 = async (str: string): Promise<string> => {
  const strArr = new TextEncoder().encode(str);
  const strDigest = await crypto.subtle.digest("SHA-256", strArr);
  const hex = Array.from(new Uint8Array(strDigest), (e) =>
    e.toString(16).padStart(2, "0"),
  ).join("");
  return hex;
};

export class LocalStorage implements StorageProviderDTO {
  private constructor(readonly storageLocation: URL) {}
  async init() {
    await fs.mkdir(this.storageLocation, { recursive: true });
  }
  async set(key: string, value: string): Promise<void> {
    const keyHex = await sha256(key);
    const location = new URL(keyHex, this.storageLocation);
    await fs.writeFile(location, value);
  }
  async get(key: string): Promise<string | undefined> {
    const keyHex = await sha256(key);
    const location = new URL(keyHex, this.storageLocation);
    if (!(await fs.exists(location))) return undefined;
    return await fs.readFile(location, "utf-8");
  }
  async has(key: string): Promise<boolean> {
    const keyHex = await sha256(key);
    const location = new URL(keyHex, this.storageLocation);
    return await fs.exists(location);
  }
  async delete(key: string): Promise<boolean> {
    const keyHex = await sha256(key);
    const location = new URL(keyHex, this.storageLocation);
    await fs.unlink(location);
    return true;
  }
  static async create(storageLocation: URL) {
    const localStorage = new LocalStorage(storageLocation);
    await localStorage.init();
    return localStorage;
  }
}
