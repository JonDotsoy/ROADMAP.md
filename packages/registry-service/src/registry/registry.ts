import type { ManifestData } from "./dtos/data.dto";
import type { StorageProviderDTO } from "./dtos/storage-provider-dto";

export class Registry {
  constructor(private storage: StorageProviderDTO) {}

  private async getMainManifest(
    owner: string,
    project_id: string,
    default_branch?: string,
  ): Promise<ManifestData> {
    const key = `${owner}/${project_id}/main`.toLowerCase();
    return (
      JSON.parse((await this.storage.get(key)) ?? "null") ?? {
        owner,
        project_id,
        default_branch: default_branch,
        branches: {},
      }
    );
  }

  private async updateManifest(
    owner: string,
    project_id: string,
    manifest: ManifestData,
  ) {
    const key = `${owner}/${project_id}/main`.toLowerCase();
    await this.storage.set(key, JSON.stringify(manifest));
  }

  async getRoadmap(owner: string, project_id: string) {
    const manifest = await this.getMainManifest(owner, project_id);

    const data = {
      manifest,
      branches: Object.fromEntries(
        await Array.fromAsync(
          Object.entries(manifest.branches),
          async ([branch, { latest }]) => {
            return [
              branch,
              await this.getRoadmapBranch(owner, project_id, latest, branch),
            ];
          },
        ),
      ),
    };

    return data;
  }

  async getRoadmapBranch(
    owner: string,
    project_id: string,
    version: string,
    branch: string,
  ) {
    const roadmapKey =
      `${owner}/${project_id}/versions/${version}/${branch}`.toLowerCase();
    return JSON.parse((await this.storage.get(roadmapKey)) ?? "null");
  }

  async updateRoadmap(
    owner: string,
    project_id: string,
    version: string,
    branch: string,
    roadmap: any,
  ) {
    const roadmapKey =
      `${owner}/${project_id}/versions/${version}/${branch}`.toLowerCase();
    const createdAt = Date.now();
    if (await this.storage.has(roadmapKey))
      throw new Error("Roadmap already exists");

    const manifest = await this.getMainManifest(owner, project_id, branch);

    manifest.branches[branch] ??= {
      latest: "-",
      versions: [],
    };

    manifest.branches[branch].versions.push(version);
    manifest.branches[branch].latest = version;

    await this.storage.set(
      roadmapKey,
      JSON.stringify({
        owner,
        project_id,
        version,
        branch,
        created_at: createdAt,
        roadmap_data: roadmap,
      }),
    );

    await this.updateManifest(owner, project_id, manifest);
  }
}
