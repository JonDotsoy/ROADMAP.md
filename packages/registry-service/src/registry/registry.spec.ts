import {
  afterAll,
  beforeAll,
  describe,
  expect,
  it,
  mock,
  setSystemTime,
} from "bun:test";
import { Registry } from "./registry";
import { LocalStorage } from "./provider/local-storage/local-storage";
import { rm, stat } from "fs/promises";
import { readdir } from "fs/promises";

const snapDir = async (dir: URL) => {
  for (const e of await readdir(dir)) {
    const fullPath = new URL(e, dir);
    const s = await stat(fullPath);
    if (s.isFile()) {
      expect(await Bun.file(fullPath).text()).toMatchSnapshot(e);
    }
  }
};

describe("Registry", () => {
  beforeAll(() => {
    setSystemTime(1000);
  });
  afterAll(() => {
    setSystemTime();
  });

  it("updates roadmap and snapshots correctly", async () => {
    const w = new URL("__workspace__/1/", import.meta.url);
    await rm(w, { recursive: true });
    const localStorage = await LocalStorage.create(w);
    const registry = await new Registry(localStorage);
    await registry.updateRoadmap("me", "foo", "1.0.0", "main", { foo: "bar" });
    await registry.updateRoadmap("me", "foo", "1.0.1", "main", { foo: "bar" });
    await registry.updateRoadmap("me", "foo", "1.0.2", "main", { foo: "bar" });
    await registry.updateRoadmap("cat", "foo", "1.0.2", "main", { foo: "bar" });
    await registry.updateRoadmap("cat", "taz", "1.0.2", "main", { foo: "bar" });
    await registry.updateRoadmap("cat", "tal", "1.0.2", "main", { foo: "bar" });
    await registry.updateRoadmap("cat", "tal", "1.0.3", "main", { foo: "bar" });
    await snapDir(w);
  });

  it("gets roadmap and snapshots correctly", async () => {
    const w = new URL("__workspace__/2/", import.meta.url);
    await rm(w, { recursive: true });
    const localStorage = await LocalStorage.create(w);
    const registry = await new Registry(localStorage);
    await registry.updateRoadmap("me", "foo", "1.0.0", "main", { foo: "bar" });
    await registry.updateRoadmap("me", "foo", "1.0.1", "main", { foo: "bar" });
    await registry.updateRoadmap("me", "foo", "1.0.2", "main", { foo: "bar" });
    await registry.updateRoadmap("me", "foo", "1.0.0", "feature/foo", {
      foo: "bar",
    });
    await registry.updateRoadmap("me", "foo", "1.0.0", "feature/taz", {
      foo: "bar",
    });

    const roadmap = await registry.getRoadmap("me", "foo");
    expect(roadmap).toMatchSnapshot();
  });
});
