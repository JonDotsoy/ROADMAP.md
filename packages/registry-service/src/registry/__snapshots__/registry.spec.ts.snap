// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Registry updates roadmap and snapshots correctly: 2e3ac51e1af17952d98bc2513b5a1b9c5b37686eca8197ebf3faba1754c9774a 1`] = `"{"owner":"cat","project_id":"foo","default_branch":"main","branches":{"main":{"latest":"1.0.2","versions":["1.0.2"]}}}"`;

exports[`Registry updates roadmap and snapshots correctly: ed00797b49c7f9c509e57bb53de0e3251ced23e31677b01121fee5681ed17100 1`] = `"{"owner":"me","project_id":"foo","version":"1.0.2","branch":"main","created_at":1000,"roadmap_data":{"foo":"bar"}}"`;

exports[`Registry updates roadmap and snapshots correctly: a3eb8a2fa06af14ad819a5eae0069e1024af71887d5f0261f8b79f441f97c8c2 1`] = `"{"owner":"cat","project_id":"tal","default_branch":"main","branches":{"main":{"latest":"1.0.3","versions":["1.0.2","1.0.3"]}}}"`;

exports[`Registry updates roadmap and snapshots correctly: b9349519e6935136539caa3c0373baf73d7f2e9771a4ffec80595390c55115f3 1`] = `"{"owner":"cat","project_id":"taz","version":"1.0.2","branch":"main","created_at":1000,"roadmap_data":{"foo":"bar"}}"`;

exports[`Registry updates roadmap and snapshots correctly: 7d8450f3cb29c7a701b430ae5cb89e04b7896cffbaaedf098182ca84f718a5b4 1`] = `"{"owner":"me","project_id":"foo","default_branch":"main","branches":{"main":{"latest":"1.0.2","versions":["1.0.0","1.0.1","1.0.2"]}}}"`;

exports[`Registry updates roadmap and snapshots correctly: db2215658f8d7fbef7642a6a716976e576f2e84929b9dbecdb313b4a663de463 1`] = `"{"owner":"me","project_id":"foo","version":"1.0.0","branch":"main","created_at":1000,"roadmap_data":{"foo":"bar"}}"`;

exports[`Registry updates roadmap and snapshots correctly: 67e28d641f86882d62f3f1f2e9fa4cb1b70b19f191b070eeaede4c4ca6f055e8 1`] = `"{"owner":"cat","project_id":"taz","default_branch":"main","branches":{"main":{"latest":"1.0.2","versions":["1.0.2"]}}}"`;

exports[`Registry updates roadmap and snapshots correctly: 2cb0edbca80e26e3fe36125ae88214d0b4766d36802016d3b0576eb276d55c95 1`] = `"{"owner":"cat","project_id":"tal","version":"1.0.3","branch":"main","created_at":1000,"roadmap_data":{"foo":"bar"}}"`;

exports[`Registry updates roadmap and snapshots correctly: 1336523d403775ef5f117916d14a09476775b46d8637059d8714dd2d05832813 1`] = `"{"owner":"me","project_id":"foo","version":"1.0.1","branch":"main","created_at":1000,"roadmap_data":{"foo":"bar"}}"`;

exports[`Registry updates roadmap and snapshots correctly: 81f207974821c36ff0ac44872580206b804c7a852905d5b6727c856293b7771c 1`] = `"{"owner":"cat","project_id":"tal","version":"1.0.2","branch":"main","created_at":1000,"roadmap_data":{"foo":"bar"}}"`;

exports[`Registry updates roadmap and snapshots correctly: e54fb06489ad0617553d1050cf1bd671928b3926285eec8c55e9719cccdd503b 1`] = `"{"owner":"cat","project_id":"foo","version":"1.0.2","branch":"main","created_at":1000,"roadmap_data":{"foo":"bar"}}"`;

exports[`Registry  1`] = `
{
  "branches": {
    "feature/foo": {
      "branch": "feature/foo",
      "created_at": 1000,
      "owner": "me",
      "project_id": "foo",
      "roadmap_data": {
        "foo": "bar",
      },
      "version": "1.0.0",
    },
    "feature/taz": {
      "branch": "feature/taz",
      "created_at": 1000,
      "owner": "me",
      "project_id": "foo",
      "roadmap_data": {
        "foo": "bar",
      },
      "version": "1.0.0",
    },
    "main": {
      "branch": "main",
      "created_at": 1000,
      "owner": "me",
      "project_id": "foo",
      "roadmap_data": {
        "foo": "bar",
      },
      "version": "1.0.2",
    },
  },
  "manifest": {
    "branches": {
      "feature/foo": {
        "latest": "1.0.0",
        "versions": [
          "1.0.0",
        ],
      },
      "feature/taz": {
        "latest": "1.0.0",
        "versions": [
          "1.0.0",
        ],
      },
      "main": {
        "latest": "1.0.2",
        "versions": [
          "1.0.0",
          "1.0.1",
          "1.0.2",
        ],
      },
    },
    "default_branch": "main",
    "owner": "me",
    "project_id": "foo",
  },
}
`;
